catkit/hub/ase_tools/gas_phase_references.py

```diff
@@ -0,0 +1,392 @@
+    for symbol in symbols:
+        added_symbols.append(symbol)
+        for candidate in candidates:
+            symbols = ase.atoms.string2symbols(candidate)
+
+            if set(added_symbols) == set(list(references.keys()) + symbols):
+                references[symbol] = candidate
+                break
+        else:
+            raise UserWarning(
+                "No candidate satisfied {symbol}. Add more candidates".format(
+                    **locals()))
+
+    return sorted(references.items(),
+                  key=lambda _x: ase.data.atomic_numbers[_x[0]]
+                  )
+
+
+def get_atomic_stoichiometry(references):
+    """
+    Given a list of references (tuples of (symbol, molecule))
+    return stoichiometry matrix that connects atomic symbols
+    to its molecular references.
+    """
+
+    n = len(references)
+    stoichiometry = np.ndarray((n, n), float)
+    stoichiometry[:] = 0.
+    key_index = {}
+    for i, (key, species) in enumerate(references):
+
+        # in case species uses a suffix like _gas
+        species = species.split('_')[0]
+
+        key_index[key] = i
+        composition = ase.atoms.string2symbols(species)
+        for j, symbol in enumerate(composition):
+            if symbol == key:
+                stoichiometry[i, i] += 1
+            else:
+                stoichiometry[i, key_index[symbol]] -= 1
+    for i in range(n):
+        stoichiometry[i, i] = 1. / stoichiometry[i, i]
+
+    for i in range(n):
+        for j in range(n):
+            if i != j:
+                stoichiometry[i, j] *= stoichiometry[j, j]
+
+    return stoichiometry.tolist()
+
+
+def get_stoichiometry_factors(adsorbates, references):
+    """
+    Take a list of adsorabtes and a corresponding reference
+    system and return a list of dictionaries encoding the
+    stoichiometry factors converting between adsorbates and
+    reference molecules.
+    """
+
+    stoichiometry = get_atomic_stoichiometry(references)
+    stoichiometry_factors = {}
+    for adsorbate in adsorbates:
+        for symbol in ase.atoms.string2symbols(adsorbate):
+            symbol_index = list(
+                map(lambda _x: _x[0], references)).index(symbol)
+
+            for (factor, (ref_symbol, ref_molecule)) in zip(
+                    stoichiometry[symbol_index], references):
+                stoichiometry_factors.setdefault(
+                    adsorbate,
+                    {})[ref_molecule] = stoichiometry_factors.setdefault(
+                    adsorbate,
+                    {}).get(
+                    ref_molecule,
+                    0) + factor
+
+        nonzero_factors = {}
+        for key, value in stoichiometry_factors[adsorbate].items():
+            if not np.isclose(value, 0.):
+                nonzero_factors[key] = value
+        stoichiometry_factors[adsorbate] = nonzero_factors
+
+    return stoichiometry_factors
+
+
+if __name__ == '__main__':
+
+    # store previously create test results
```

Comment:
Consider storing such long dictionaries as JSON data files

Reply:
@mhoffman - can you look into this?